gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,138 @@
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+
+class UDPClient:
+    """Client for sending UDP datagrams.
+
+    Args:
+        ip (str): Destination IP address.
+        port (int): Destination port.
+        loop (asyncio.AbstractEventLoop impl): (optional) Event loop.
+    """
+    def __init__(self, ip, port, loop=None):
+        self.ip = ip
+        self.port = port
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: FfwdClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """ITimer implementation which sends UDP messages to FFWD on completion.
+
+    Args:
+        metric (dict): Dict representation of the metric to send.
+        udp_client (UDPClient): A metric sending client.
+        time_unit (number): (optional) Scale time unit for use with time.time()
+            (example: 1E9 to send nanoseconds).
+    """
+    def __init__(self, metric, udp_client, time_unit=1):
+        self.metric = metric
+        self.udp_client = udp_client
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        self._start_time = time.time()
+
+    async def stop(self):
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        await self.udp_client.send(self.metric)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class FfwdRelay:
+    """IMetricRelay implementation which sends metrics to FFWD immediately.
+
+    Args:
+        config (dict): Required keys:
+            key (str): A key to identify the service as a whole.
+            ffwd_ip (str): ffwd IP address.
+            ffwd_port (int): ffwd port number.
+            time_unit (number): Scale unit to use with timers.
+    """
+    def __init__(self, config):
+        self.key = config['key']
+        self.udp_client = UDPClient(config['ffwd_ip'], config['ffwd_port'])
+        self.time_unit = config['time_unit']
+
+    def _create_metric(self, metric_name, value, **kwargs):
+        attrs = kwargs.get('attrs') or {}
+        tags = kwargs.get('tags') or []
+
+        attrs['what'] = metric_name
+        return {
+            'key': self.key,
+            'attributes': attrs,
+            'value': value,
+            'type': 'metric',
+            'tags': tags
+        }
+
+    async def incr(self, metric_name, value=1, **kwargs):
+        await self.set(metric_name, value, **kwargs)
```

Comment:
Well, our increment is actually server-side, so we just emit `1` a bunch of times and do the sum in the Alien graph. Hmm maybe this should be called AlienFfwdRelay or at least include that in one of the docstrings?  @matt0bi Thoughts?

Reply:
>  Hmm maybe this should be called AlienFfwdRelay or at least include that in one of the docstrings?

To clarify, FFWD + Heroic can be used with front-ends different than Alien, but in this case we're completely relying on Alien to do the accumulation for us? I know at least one other tsdb that uses the same approach (influxdb and its derivative fn), so I think including something in the docstring about this should be fine.

Reply:
I was also thinking alien was too specific for a name, but I don't think specifying a bit more would be bad. How about something like `FastFfwdRelay` or some better name (`NonAggregatingFfwdRelay` seems too long)?  The thing I'm trying to clarify from the name is to differentiate the metrics are sent immediately, since it would be perfectly valid to set up a ffwd implementation that had a layer of aggregation in front of it (i.e. has an internal counter, instead of sending only 1's).

Reply:
What about `SimpleFfwdRelay`? I think `FastFfwdRelay` is good. I think that an `AggregatingFfwdRelay` is a special case which needs explicit clarification whereas a non-aggregating relay is the base case. Just to be clear, I think either updating the name to `FastFfwdRelay` or adding something to the docstring would work fine here.

Reply:
+1 to SimpleFfwdRelay and Aggregating